geoSegment.js

```diff
@@ -0,0 +1,353 @@
+
+const SESSION_SERIES_FILE_PATHS = [];
+
+// Event Schedule generation
+const SCHEDULE_TYPE = 'Schedule';
+const MINIMAL_EVENT_SCHEDULE_SCHEMA = Joi.object().keys({
+  // If this array has length 0, an infinite loop will occur in the date
+  // generation algorithm below.
+  byDay: Joi.array().items(Joi.string()).min(1).required(),
+  startTime: Joi.string().required(),
+  endTime: Joi.string().required(),
+  idTemplate: Joi.string().required(),
+});
+
+/**
+ * @typedef {import('axios').AxiosError} AxiosError
+ * @typedef {import('axios').AxiosRequestConfig} AxiosRequestConfig
+ * @typedef {import('axios').AxiosResponse} AxiosResponse
+ * @typedef {{
+ * location?: {
+ *  geo?: {
+ *   latitude: number,
+ *   longitude: number,
+ *   }
+ *  },
+ * [k:string]: unknown
+ * }} SessionSeriesData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'SessionSeries',
+ *  data: SessionSeriesData
+ * }} SessionSeriesItem
+ * @typedef {{
+ *  startDate?: string,
+ *  superEvent?: string,
+ *  [k:string]: unknown
+ * }} ScheduledSessionData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'ScheduledSession',
+ *  data: ScheduledSessionData
+ * }} ScheduledSessionItem
+ * @typedef {{
+ *  latitude: number,
+ *  longitude: number,
+ *  radius: number,
+ *  identifier: string
+ * }} Segment
+ */
+
+/**
+ * @param {Segment[]} segments
+ */
+async function createSegmentDirectories(segments) {
+  // Create segment directories;
+  segments.map((segment) => segment.identifier)
+    .map(async (segmentIdentifier) => {
+      const segmentDirectoryPath = `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}`;
+      await emptyOrMakeDirectory(segmentDirectoryPath);
+
+      // Create index file
+      await fsPromises.writeFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, '');
+    });
+}
+
+/**
+ * @param {ScheduledSessionData} scheduledSessionData
+ * @param {SessionSeriesData} sessionSeries
+ */
+async function linkScheduledSessionAndSessionSeriesAndWrite(scheduledSessionData, sessionSeries) {
+  // Link ScheduledSession with SessionSeries
+  const linkedScheduledSessionData = {
+    ...scheduledSessionData,
+    superEvent: sessionSeries,
+  };
+
+  // Write ScheduledSession into each output segment directory
+  for (const segmentIdentifier of sessionSeries['imin:segment']) {
+    const filename = `${encodeURIComponent(linkedScheduledSessionData.id)}.json`;
+    await fs.writeJson(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/${filename}`, linkedScheduledSessionData);
+    await fs.appendFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, `${filename}\r\n`);
+  }
+}
+
+/**
+ * There is a system filepath limit of 255 characters in most file systems.
+ * All sessionseries filepaths with be of the form `./sessionseries/${encodedFilename}.json`
+ * Therefore the encodedFilename has to be less than 234 characters in length (./sessionseries/.json is 21 characters)
```

Comment:
This is clever! However, is the file path limit supposed to refer to the whole absolute path rather than the relative path? So if this is starting in a directory that is at `/home/ilikereallylongusernames/projects/openobjects/version/number/thirteen/` or something, then that would change the calculation. You could calculate this current directory length by using `__dirname`.
I would also suggest calculating the 21 using the `SESSION_SERIES_DIRECTORY_PATH.length` just in case they (or we) change this name in future. They've modified our scripts before.

One other suggestion to limit the chance of long file paths is: could it be a hash on the ID? Not sure if that breaks anything else

Reply:
I was just reading this and wondering about a hash also... perhaps that would be a safer bet here in fact, as this filename limit could indeed cause issues - using a hash of the ID for both SessionSeries and ScheduledSession filenames

To account for the (unlikely) case of hash collision, we should:
- For SessionSeries: Double-check the `@id` within the SessionSeries matches the one we expect when we retrieve the SessionSeries from the hash file, and ignore it if not (logging a collision error). This means we'll drop records rather than write erroneous data.
- For ScheduledSession: Ensure (as commented elsewhere) that we check if the file exists before updating index.txt with the ScheduledSession filename, hence again we'll drop data if a collision occurs (as it's overwriting an existing record). If the file already exists we should read it to check the `@id`, and again if there's a mismatch log a collision error.

This way a test run against production firehose will give us a clear indication of if the hashing algorithm we've chosen is creating too many collisions.

Reply:
Yes, hashing is a good way to go! Have implemented